### PR TITLE
Advanced Search - API

### PR DESCRIPTION
**Pull request by wooleyHeadspring**
This is a manual merge from [v0.4.30-beta](https://github.com/kleinisd-lp/Leadership-Profile/releases/tag/v0.4.30-beta) in the Klein repo.  Changes were manual to ensure the commits are properly signed.  This PR includes the addition of views, controllers, models, and tests that add the "Advanced Search" functionality.

The tests included are not validating the data in the database, but rather that the contracts and queries are functionally sound.

### CODE REVIEW

File: src/API/LeadershipProfileAPI/Startup.cs [hunk 1]
```diff
@@ -131,7 +131,7 @@ private static void AddAuth(IServiceCollection services)
                 // this defines a CORS policy called "default"
                 options.AddPolicy("default", policy =>
                 {
-                    policy.WithOrigins(authorityServer, webClient)
+                    policy.WithOrigins(new string[] { authorityServer, webClient, "http://localhost", "https://localhost" })
```
Comment (YogirajA) [inline src/API/LeadershipProfileAPI/Startup.cs:134]: I wonder why http is necessary.

**Review by YogirajA — CHANGES_REQUESTED** [review]
Besides the inline comments, please address below

- [x] Fix the failing build
- [ ] Search from swagger is giving **Conversion failed when converting date and/or time from character string** error. It needs to be fixed.

File: src/API/LeadershipProfileAPI/Controllers/ClaimCheckAttribute.cs [hunk 2]
```diff
@@ -11,7 +11,7 @@ public class ClaimCheckAttribute : TypeFilterAttribute
     {
         public ClaimCheckAttribute(string claimType, string claimValue) : base(typeof(ClaimCheckFilter))
         {
-            Arguments = new object[] {claimType , claimValue};
+            Arguments = new object[] { claimType, claimValue };
```
[the remark this answers is not in the document]
Comment (wooleyHeadspring) [reply src/API/LeadershipProfileAPI/Controllers/ClaimCheckAttribute.cs:14]: Duly noted.  Will keep that in mind in the future.

File: src/API/LeadershipProfileAPI/Data/Models/ListItem/ListItemCategory.cs [hunk 3]
```diff
@@ -0,0 +1,8 @@
+﻿namespace LeadershipProfileAPI.Data.Models.ListItem
+{
+    public class ListItemCategory
```
[the remark this answers is not in the document]
Comment (wooleyHeadspring) [reply src/API/LeadershipProfileAPI/Data/Models/ListItem/ListItemCategory.cs:3]: Good idea. Will do that.

File: src/API/LeadershipProfileAPI/Data/Models/ListItem/ListItemSubCategory.cs [hunk 4]
```diff
@@ -0,0 +1,9 @@
+﻿namespace LeadershipProfileAPI.Data.Models.ListItem
+{
+    public class ListItemSubCategory
```
[the remark this answers is not in the document]
Comment (wooleyHeadspring) [reply src/API/LeadershipProfileAPI/Data/Models/ListItem/ListItemSubCategory.cs:3]: ListItemSubCategory will now derive from the same ListItemBsae

File: src/API/LeadershipProfileAPI/Extensions/ListExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+        /// <summary>
+        /// Extension method to add a string to a List of string types if the value is not null nor whitespace
+        /// </summary>
+        /// <param name="list">The list the value should be added to</param>
+        /// <param name="value">The value to add to the list</param>
+        public static void AddIfNotNullOrWhiteSpace(this List<string> list, string value)
+        {
+            if (list != null)
+            {
+                if (!string.IsNullOrWhiteSpace(value))
```
[the remark this answers is not in the document]
Comment (wooleyHeadspring) [reply src/API/LeadershipProfileAPI/Extensions/ListExtensions.cs:22]: This extension class will be removed as it is no longer used after cleanup of EdFiDbQueryData.GetSearchResultsAsync.

File: src/API/LeadershipProfileAPI/Startup.cs (hunk 1 above)
Comment (wooleyHeadspring) [reply src/API/LeadershipProfileAPI/Startup.cs:134]: The Client defaults to port 80 and there is no "HTTPS=true" on the startup scripts in [package.json](https://github.com/Ed-Fi-Exchange-OSS/Leadership-Profile/blob/e9f988db1327e3db7f3cc15134c9b737ad243603/src/Web/package.json#L23)

**Comment by wooleyHeadspring** [comment]
@YogirajA, comments have been addressed from feedback and committed [here](https://github.com/Ed-Fi-Exchange-OSS/Leadership-Profile/pull/51/commits/9e505cb1fc201f8ae15c1c342776fb6fd6b725ff)

As for "Search from swagger is giving Conversion failed when converting date and/or time from character string error. It needs to be fixed.". This was addressed by changing the data type for the properties to DateTime from string.  This will throw an error, but it is much more specific on what the issue is...
{
  "type": "https://tools.ietf.org/html/rfc7231#section-6.5.1",
  "title": "One or more validation errors occurred.",
  "status": 400,
  "traceId": "00-74c4c8cb7ea68f48aae759a001ef1e6b-606aa7c22b3e204b-00",
  "errors": {
    "$.certifications.issueDate": [
      "The JSON value could not be converted to System.Nullable`1[System.DateTime]. Path: $.certifications.issueDate | LineNumber: 0 | BytePositionInLine: 125."
    ]
  }
}

**Review by YogirajA — APPROVED** [review]

**Pull request merged**